lib/expression/eval.ex

```diff
@@ -70,9 +69,8 @@ defmodule Expression.Eval do
   def eval!({:function, opts}, context, mod) do
     function_name = opts[:name] || raise "Functions need a name"
     args = opts[:args] || []
-    arguments = Enum.reduce_while(args, [], &args_reducer(&1, function_name, context, mod, &2))
```

Comment:
It should also make the library a lot faster because now it can short-circuit

Reply:
it's interesting as the intention of using reduce while here was to only evaluate until one of the things resolved to true. I'm happy that it's now solved but I'm unsure _why_ it's solved. I'm going to need to spend some more time on this to make sure I understand.

Reply:
It seems like the only reason to do this here is: 

- avoid issues with the "or" but I also don't understand why fixing it here 
- avoid having to pass the custom callback as argument of every function both in the standard lib and custom callback module functions

Reply:
Given that the custom callback module does not change between function calls, I didn't see it as problematic to add it here and so pass it along every call to `eval!` in the `Expression.Callbacks.EvalHelpers`. Just like we store temporary values in the context with `__value__`, `__captures`, `__type__` etc. 

It is basically the implementation of the state monad